### PR TITLE
Adds support for in-server service name resolution.

### PR DESCRIPTION
**Pull request by filmil**
Note: the full design proposal is available at: https://goo.gl/APfsCE

This change allows webhook authorizer that call out to an endpoint of
the form:

    https://servicename.namespace.svc/someurl

to resolve "servicename.namespace.svc" to the service's current
endpoint, without contacting kube-dns.

This allows deployments without kube-dns access (e.g. GKE, or the
default k8s-on-GCP) to host the webhook authorizer as a pod.  It is
somewhat similar to the resolution already used for say admission
control plugins, but has a few important differences:

- The configuration files for admission control and webhook authz are
different.

- There is no dynamic registration support for authz webhooks.

The feature depends on a recent change that requires server certificate
to have CN subject field in the certificate be set to the Kubernetes
service name (for example: "/CN=servicename.namespace.svc").

Also, since the authz webhook is already released and being used, we
must uphold a few constraints:

- Deployments that don't need or use this feature should not have to pay
the cost of having it around.

- Since kube-apiserver is the only server that would need this feature,
the changes are confined to kube-apiserver specific code, and not pushed
deeper into generic webhook.

Changes inside:

- Introduced a flag --authorization-webhook-use-service-resolution
(default: false) that gates the service resolution.  This is because
there may be deployments that can do service resolution from the master
node through kube-dns, which do not need this additional functionality.
Those users can simply keep the flag set to its default value and
pretend that this functionality does not exist.

- Wired through a service resolver and the proxy transport dialer from
the apiserver setup code into authorizer.  This is a very similar
approach to how this is done today with the admission control code.

- In the authz webhook proper, added a whole optional section that
periodically recreates the webhook client based off of the freshest
available endpoint IP address for the service in question.  The optional
section only makes a difference when the resolution is used.  Otherwise,
the code paths are not extended.  I did introduce a lock for the client
refresh; but I expect that the contention would be significant in
regular use, since write locks are rare, and regular requests use the
read locks.

- Some refactoring and reformatting of the existing code.  Namely the
webhook client construction was broken up into two smaller pieces:
reading the kubeconfig file, and creating the client proper from the
kubeconfig file.

**What this PR does / why we need it**: This pull request handles item (2) from https://github.com/kubernetes/kubernetes/issues/52511  It mirrors the approach that is already in use
for webhook admission control for example, but is adjusted to fit neatly in the current kubeconfig-based
webhook authz configuration.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Adds support for in-server service name resolution.

This change allows webhook authorizer that call out to an endpoint of
the form:

    https://servicename.namespace.svc/someurl

to resolve "servicename.namespace.svc" to the service's current
endpoint, without contacting kube-dns.

Design proposal is available at: https://goo.gl/APfsCE
```

### CODE REVIEW

**Comment by cmluciano** [comment]
@kubernetes/sig-network-pr-reviews

**Comment by liggitt** [comment]
thanks for the PR. this is related to the discussion in https://github.com/kubernetes/kubernetes/issues/52511

File: staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go [hunk 1]
```diff
@@ -58,35 +90,61 @@ func NewFromInterface(subjectAccessReview authorizationclient.SubjectAccessRevie
...
-// For additional HTTP configuration, refer to the kubeconfig documentation
-// https://kubernetes.io/docs/user-guide/kubeconfig-file/.
-func New(kubeConfigFile string, authorizedTTL, unauthorizedTTL time.Duration) (*WebhookAuthorizer, error) {
-	subjectAccessReview, err := subjectAccessReviewInterfaceFromKubeconfig(kubeConfigFile)
+//         # Certificate for the webhook plugin to use when calling out to the
+//         # webhook server.
+//         client-certificate: /path/to/cert.pem
+//         # The client's private key that signed the certificate above.
+//         client-key: /path/to/key.pem
+func New(
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go]: the webhook authorizer is used by multiple clients, other than just the main kube-apiserver (which is why it lives under `k8s.io/apiserver`, not `k8s.io/kubeapiserver`).

only kube-apiserver can be expected to have the ability, or the need (and only on some deployment topologies) to do manual resolution of service names to IPs.

there is already a `NewFromInterface` constructor that allows passing in a client that submits subjectaccessreviews opaquely. If kube-apiserver wanted to use non-dns resolution to dial service IPs, it should construct a client to do that and use that constructor, rather than make the webhook authorizer aware of service resolution like this

File: staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go [hunk 2]
```diff
@@ -33,23 +38,50 @@ import (
...
 )
 
-const retryBackoff = 500 * time.Millisecond
+const (
+	retryBackoff = 500 * time.Millisecond
+
+	// The default webhook client lifetime, in case service resolution is
+	// used.  The webhook client is recreated after it lives for at least
+	// this long.
+	webhookClientTTL = 1 * time.Minute
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go]: I wouldn't expect any changes inside the `k8s.io/apiserver` package

File: staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go (hunk 1 above)
Comment (filmil) [reply staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go]: Thanks for the comments.  I'll wait a bit to give anyone else opportunity to chime in.  Then will go implement this change.

**Comment by liggitt** [comment]
cc @kubernetes/sig-auth-proposals @kubernetes/sig-api-machinery-proposals

File: pkg/kubeapiserver/options/authorization.go [hunk 3]
```diff
@@ -60,6 +66,12 @@ func (s *BuiltInAuthorizationOptions) AddFlags(fs *pflag.FlagSet) {
 		"File with webhook configuration in kubeconfig format, used with --authorization-mode=Webhook. "+
 		"The API server will query the remote service to determine access on the API server's secure port.")
 
+	fs.BoolVar(&s.WebhookUseServiceResolution, "authorization-webhook-use-service-resolution", s.WebhookUseServiceResolution,
+		"If set, the authorizer webhook code will resolve URLs pointing to hosts of the form "+
+			"https://servicename.namespace.svc:port/somewhere to appropriate endpoint IP addresses. "+
+			"You may want to use this feature if your webhook authorizer is hosted in the cluster itself, "+
+			"and the apiserver can not access kube-dns.")
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authorization.go]: Phrase this more generically... `kube-dns` is only one possible in-cluster DNS implementation

**Review by filmil — COMMENTED** [review]
I believe that I handled the remarks on the review. 

I pulled the subject access review client up into kube-apiserver.

**Comment by deads2k** [comment]
I think you could encapsulate the `ServiceResolver` concept into the `Dial` func of a `rest.Config` to limit the ripples and use our existing mechanisms.  See the func here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L118 .  You have the address, you can parse it there to figure out what you'd like to do.

**Comment by filmil** [comment]
On Fri, Oct 20, 2017 at 9:46 AM David Eads <notifications@github.com> wrote:

> I think you could encapsulate the ServiceResolver concept into the Dial
> func of a rest.Config to limit the ripples and use our existing
> mechanisms. See the func here:
> https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L118
> . You have the address, you can parse it there to figure out what you'd
> like to do.
>
What happens if the underlying service IP address changes?  Is the
client-go smart enough to attempt to reconnect to a different IP address in
that case?

Two main reasons I decided to stop at the ServiceResolver level:
(1) It's already an existing concept used for service resolution in
admission control for example.  Having a choice, the code there also uses a
separate concept of a ServiceResolver and recreates a client on demand.
This strongly suggested to me that this is the desired approach for service
resolution.

(2) It was not clear to me from reading the docs that Dial would have been
called more than once in the lifetime of a client.  It depends a lot on how
the client then behaves in case of intermittent connection errors, and it
is not clear to me that one can recover easily from within a Dial function.

F

>

**Comment by liggitt** [comment]
> I think you could encapsulate the ServiceResolver concept into the Dial func of a rest.Config to limit the ripples and use our existing mechanisms. See the func here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/rest/config.go#L118 . You have the address, you can parse it there to figure out what you'd like to do.

I like that a lot, since the kube-apiserver service resolver impls are essentially reimplementing DNS lookup, `Dial` is the right place to do that.

> What happens if the underlying service IP address changes? Is the client-go smart enough to attempt to reconnect to a different IP address in that case?

The client deals in service DNS names. The next time the connection is established, the dialer can re-resolve to a new IP and dial that.

> It was not clear to me from reading the docs that Dial would have been called more than once in the lifetime of a client. It depends a lot on how the client then behaves in case of intermittent connection errors

Dial is called for every net.Conn establishment.

**Comment by filmil** [comment]
On Fri, Oct 20, 2017 at 10:18 AM Jordan Liggitt <notifications@github.com>
wrote:

> I like that a lot, since the kube-apiserver service resolver impls are
> essentially reimplementing DNS lookup, Dial is the right place to do that.
>
> What about the existing ServiceResolver concept in admission control?
Isn't it also something that could in principle be pushed down into a
Dialer?

> The client deals in service DNS names. The next time the connection is
> established, the dialer can re-resolve to a new IP and dial that.
>
> What about long-running connections?  Whose responsibility is it to
reconnect?  Would a client reattempt to connect automatically, or would the
code that uses the client need to handle retries?  The current
implementation will recover on its own, for example.

> Dial is called for every net.Conn establishment.
>
> Noted, thanks.

F

**Comment by liggitt** [comment]
> The current implementation will recover on its own, for example.

The current implementation depends on reconnection as well.

**Comment by filmil** [comment]
On Fri, Oct 20, 2017 at 10:33 AM Jordan Liggitt <notifications@github.com>
wrote:

> The current implementation will recover on its own, for example.
>
> The current implementation depends on reconnection as well.
>

That's true.  But it will try to detect if reconnect is needed and will
reconnect.

Will the rest client based on a custom Dialer also do that?

F

**Comment by liggitt** [comment]
> That's true. But it will try to detect if reconnect is needed and will reconnect.

Actually, it looks to be expiring the client based on a timeout only.

Assuming it changed to actually see if `ResolveEndpoint()` returned something different, for services backed by multiple pods, the pod-ip-resolving implementation randomly selects an endpoint, so even in cases where you already had a client with active keep-alive connections to a still-valid pod, this impl would stand a good chance of tearing them down for no reason.

> Will the rest client based on a custom Dialer also do that?

Yes, if its keep-alive connections are dropped, the next API call attempted will re-call Dial, at which point it can re-resolve the address to establish new the connection.

**Comment by filmil** [comment]
On Fri, Oct 20, 2017 at 5:53 PM Jordan Liggitt <notifications@github.com>
wrote:

> That's true. But it will try to detect if reconnect is needed and will
> reconnect.
>
> Actually, it looks to be expiring the client based on a timeout only.
>
No; it actually skips in case of no change in IP address.

// Only recreate the client if the endpoint address has actually
 + // changed since the last time we checked.
 + if resolvedUrl.String() != t.lastResolvedUrl.String() {
 + config.Host = resolvedUrl.String()
 + glog.V(6).Infof("resolved target: %+v", config.Host)
 + t.w, err = utilwebhook.NewClientFromConfig(&config, retryBackoff)
 + if err != nil {
 + return errors.Wrapf(err, "while recreating webhook client")
 + }
 + t.lastResolvedUrl = *resolvedUrl
 + }
 + t.webhookExpiresAt = time.Now().Add(webhookClientTTL)
 + return nil


> Yes, if its keep-alive connections are dropped, the next API call
> attempted will re-call Dial, at which point it can re-resolve the address
> to establish new the connection.
>
So your final recommendation is to go with Dial?

F

**Comment by liggitt** [comment]
> No; it actually skips in case of no change in IP address.

Ah, I see. That's where the random selection implementation of `ResolveEndpoint()` would have bitten you, causing your client to potentially be rebuilt on every request.

> So your final recommendation is to go with Dial?

It has the info needed, avoids rebuilding the client, you get keep-alive connection pooling for free, and the changes required are way smaller. Something like:
1. conditionally build a dialer in `kube-apiserver/app/server.go#BuildGenericConfig`

    ```
    func buildAuthorizationWebhookDialer(serviceResolver aggregatorapiserver.ServiceResolver, proxyTransport *http.Transport) func(network, address string) (net.Conn, error) {
    	baseDialer := net.Dial
    	if proxyTransport != nil && proxyTransport.Dial != nil {
    		baseDialer = proxyTransport.Dial
    	}
    
    	return func(network, address string) (net.Conn, error) {
    		host, _, err := net.SplitHostPort(address)
    		if err != nil {
    			return baseDialer(network, address)
    		}
    
    		if !strings.HasSuffix(host, ".svc") {
    			return baseDialer(network, address)
    		}
    
    		segments := strings.Split(host, ".")
    		if len(segments) != 3 {
    			return baseDialer(network, address)
    		}
    
    		u, err := serviceResolver.ResolveEndpoint(segments[1], segments[0])
    		if err != nil {
    			return nil, err
    		}
    
    		return baseDialer(network, u.Host)
    	}
    }
    ```
2. pass the dialer to `BuildAuthorizer()` -> `ToAuthorizationConfig()` -> `AuthorizationConfig{}`
3. in `AuthorizationConfig#New()`:

    ```go
    // Load config
    loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
    loadingRules.ExplicitPath = config.WebhookConfigFile
    loader := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
    clientConfig, err := loader.ClientConfig()
    if err != nil {
            return nil, nil, fmt.Errorf("invalid webhook config: %v", err)
    }
    
    // Set timeout and dialer
    clientConfig.Timeout = 30 * time.Second
    clientConfig.Dial = config.WebhookDialer
    
    // Build client
    webhookClient, err := authorizationclient.NewForConfig(clientConfig)
    if err != nil {
            return nil, nil, err
    }

    // Build authorizer
    webhookAuthorizer, err := webhook.NewFromInterface(webhookClient.SubjectAccessReviews(),
      config.WebhookCacheAuthorizedTTL,
      config.WebhookCacheUnauthorizedTTL)
    ```

**Comment by smarterclayton** [comment]
+1 Dial is the correct abstraction for establishing name resolution

File: cmd/kube-apiserver/app/server.go [hunk 4]
```diff
@@ -552,9 +554,49 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
...
+	// Following sample code from issue #54163 by liggitt@redhat.com.
+	baseDialer := net.Dial
+	if proxyTransport != nil && proxyTransport.Dial != nil {
+		baseDialer = proxyTransport.Dial
+	}
+
+	return func(network, address string) (net.Conn, error) {
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			glog.Warning("while splitting hostport: %v", err)
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: for debugging, I'm assuming... I'd recommend just passing through and letting the underlying dialer return an error about a malformed address

File: cmd/kube-apiserver/app/server.go [hunk 5]
```diff
@@ -552,9 +554,49 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
...
+			return baseDialer(network, address)
+		}
+		// "servicename.namespace.svc"
+		segments := strings.Split(host, ".")
+		if len(segments) != 3 {
+			glog.V(1).Infof("Dialer address is not a Kubernetes svc reference: %v", host)
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: same here, no logging needed

File: cmd/kube-apiserver/app/server.go [hunk 6]
```diff
@@ -552,9 +554,49 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
...
+			return baseDialer(network, address)
+		}
+		u, err := serviceResolver.ResolveEndpoint(segments[1], segments[0])
+		if err != nil {
+			return nil, errors.Wrapf(
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: huh, didn't know about Wrapf... neat.

File: cmd/kube-apiserver/app/server.go [hunk 7]
```diff
@@ -443,7 +444,8 @@ func BuildGenericConfig(s *options.ServerRunOptions, proxyTransport *http.Transp
 		return nil, nil, nil, nil, nil, fmt.Errorf("invalid authentication config: %v", err)
 	}
 
-	genericConfig.Authorizer, genericConfig.RuleResolver, err = BuildAuthorizer(s, sharedInformers)
+	genericConfig.Authorizer, genericConfig.RuleResolver, err =
+		BuildAuthorizer(s, sharedInformers, serviceResolver, proxyTransport)
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: missing dialer param, and missing conditional building of dialer using buildAuthorizationWebhookDialer, right?

File: federation/cmd/federation-apiserver/app/server.go [hunk 8]
```diff
@@ -193,7 +193,7 @@ func NonBlockingRun(s *options.ServerRunOptions, stopCh <-chan struct{}) error {
 	}
 	versionedInformers := clientgoinformers.NewSharedInformerFactory(clientgoExternalClient, 10*time.Minute)
 
-	authorizationConfig := s.Authorization.ToAuthorizationConfig(sharedInformers)
+	authorizationConfig := s.Authorization.ToAuthorizationConfig(sharedInformers, nil /*serviceResolver*/, nil /*proxyTransport*/)
```
Comment (liggitt) [inline federation/cmd/federation-apiserver/app/server.go]: switch to nil dialer param

File: staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go [hunk 9]
```diff
@@ -42,12 +42,18 @@ type GenericWebhook struct {
 	initialBackoff time.Duration
 }
 
-// NewGenericWebhook creates a new GenericWebhook from the provided kubeconfig file.
-func NewGenericWebhook(registry *registered.APIRegistrationManager, codecFactory serializer.CodecFactory, kubeConfigFile string, groupVersions []schema.GroupVersion, initialBackoff time.Duration) (*GenericWebhook, error) {
-	return newGenericWebhook(registry, codecFactory, kubeConfigFile, groupVersions, initialBackoff, defaultRequestTimeout)
-}
-
-func newGenericWebhook(registry *registered.APIRegistrationManager, codecFactory serializer.CodecFactory, kubeConfigFile string, groupVersions []schema.GroupVersion, initialBackoff, requestTimeout time.Duration) (*GenericWebhook, error) {
+// LoadClientConfig loads a kubeconfig-formatted file at filesystem path
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/util/webhook/webhook.go]: let's split unrelated util/webhook and authorizer/webhook changes out of this PR (and I think just the doc changes are relevant now)

**Comment by filmil** [comment]
Thanks. Please stand by, II'm working the suggested changes into the PR.

File: cmd/kube-apiserver/app/server.go (hunk 4 above)
Comment (filmil) [reply cmd/kube-apiserver/app/server.go]: Changed to a verbose log message. If you don't mind, I'd leave a verbose log here.  While yes, passing through will eventually surface an error message, it gets very difficult to figure out the error message source if you are not very familiar with the code path.  This log will reveal at least part of it.  This has, at least to me, been useful numerous times in debugging.

File: cmd/kube-apiserver/app/server.go (hunk 7 above)
Comment (filmil) [reply cmd/kube-apiserver/app/server.go]: I think you were looking at an intermediate phase of the code as I was reworking it. Please stand by, I'll let you know when the change is ready for review.  Otherwise, you'll time looking at defunct code.

Do you get a new update when I change my new feature branch?  If so, I'll try to minimize that to remove disruption to your workflow. Sorry.

File: cmd/kube-apiserver/app/server.go (hunk 6 above)
Comment (filmil) [reply cmd/kube-apiserver/app/server.go]: Ack

**Comment by filmil** [comment]
Please take another look.

I think most of your remarks are now fixed.  I opted to leave the verbose logging messages, since they really help understand the code paths.  I would really like to keep them unless you feel very strongly about not having them there.

File: cmd/kube-apiserver/app/server.go [hunk 10]
```diff
@@ -444,7 +444,9 @@ func BuildGenericConfig(s *options.ServerRunOptions, proxyTransport *http.Transp
 		return nil, nil, nil, nil, nil, fmt.Errorf("invalid authentication config: %v", err)
 	}
 
-	genericConfig.Authorizer, genericConfig.RuleResolver, err = BuildAuthorizer(s, sharedInformers)
+	genericConfig.Authorizer, genericConfig.RuleResolver, err =
+		BuildAuthorizer(s, sharedInformers,
+			buildAuthorizationWebhookDialer(serviceResolver, proxyTransport))
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: missed making this dialing conditional/optional

**Comment by filmil** [comment]
Hm, I end up seeing the following message:

```
E1023 14:53:47.939251   21163 webhook.go:202] Failed to make webhook authorizer request: the server could not find the requested resource (post subjectaccessreviews.authorization.k8s.io)
```

Not sure what the message means, though.  Asked the same question at #sig-auth.

File: pkg/kubeapiserver/authorizer/config.go [hunk 11]
```diff
@@ -105,11 +111,32 @@ func (config AuthorizationConfig) New() (authorizer.Authorizer, authorizer.RuleR
...
+			clientConfig.Dial = config.WebhookDialer
+
+			// Build client
+			webhookClient, err := authorizationclient.NewForConfig(clientConfig)
+			if err != nil {
+				return nil, nil, fmt.Errorf("while creating webhook client: %v", err)
+			}
+
+			webhookAuthorizer, err := webhook.NewFromInterface(
+				webhookClient.SubjectAccessReviews(),
```
Comment (ericchiang) [inline pkg/kubeapiserver/authorizer/config.go]: This client automatically sets the URL path. See the existing webhook authorizer for how to create a client that sends access reviews to the root path https://github.com/kubernetes/kubernetes/blob/9b8d70dc426f7203f2a32e4ec0e4117c204fee0d/staging/src/k8s.io/apiserver/plugin/pkg/authorizer/webhook/webhook.go#L235-L245

**Comment by filmil** [comment]
PTAL.  I added some useful unit tests.

**Review by caesarxuchao — COMMENTED** [review]
Merging the service resolver into the dialer looks right.

**Comment by filmil** [comment]
On Tue, Oct 24, 2017 at 5:12 PM Chao Xu <notifications@github.com> wrote:

> *@caesarxuchao* commented on this pull request.
>
> Merging the service resolver into the dialer looks right.
> ------------------------------
>
> In cmd/kube-apiserver/app/server.go
> <https://github.com/kubernetes/kubernetes/pull/54163#discussion_r146710742>
> :
>
> > @@ -558,9 +560,56 @@ func BuildAuthenticator(s *options.ServerRunOptions, storageFactory serverstorag
>  	return authenticatorConfig.New()
>  }
>
> +// buildAuthorizationWebhookDialer constructs a custom dialer function for the authorization webhook
>
> buildWebhookDialer? Admission webhooks will reuse this function.
>

Done.


> ------------------------------
>
> In cmd/kube-apiserver/app/server.go
> <https://github.com/kubernetes/kubernetes/pull/54163#discussion_r146711221>
> :
>
> > +			glog.V(6).Infof("Dialer address is not a Kubernetes svc reference: %v", host)
> +			return baseDialer(network, address)
> +		}
> +		namespace := segments[1]
> +		service := segments[0]
> +		u, err := serviceResolver.ResolveEndpoint(namespace, service)
> +		if err != nil {
> +			return nil, fmt.Errorf(
> +				"while resolving endpoint: namespace=%v, service=%v: %v",
> +				namespace, service, err)
> +		}
> +		glog.V(5).Infof("Resolved: namespace=%v, service=%v to: host=%v",
> +			namespace, service, u.Host)
> +		return baseDialer(network, u.Host)
> +	}
> +
>
> remove the extra new line
>

Done

**Comment by filmil** [comment]
@liggitt @caesarxuchao please take another look.

All suggestions should now be included, and all presubmit tests fixed.

**Comment by filmil** [comment]
@liggitt @caesarxuchao Ping?

File: cmd/kube-apiserver/app/server_test.go [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+	for i, ttt := range tt {
+		var transport *http.Transport
+		defaultTransport := &http.Transport{}
+		if ttt.useNilTransport {
+			defaultTransport = nil
+		}
+		transport = &http.Transport{
+			Dial: buildWebhookDialer(&resolver, defaultTransport)}
+		client := &http.Client{Transport: transport}
+		response, err := client.Get(ttt.serverAddress)
```
Comment (ericchiang) [inline cmd/kube-apiserver/app/server_test.go]: Is this just making a call to the internet?

File: cmd/kube-apiserver/app/server_test.go [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+		if err != nil {
+			if ttt.expectedError == nil {
+				t.Errorf("[%v] unexpected error: %v", i, err)
+			} else if strings.Index(err.Error(), ttt.expectedError.Error()) == -1 {
```
Comment (ericchiang) [inline cmd/kube-apiserver/app/server_test.go]: If this error message isn't created your code, your test probably shouldn't depend on the exact string.

File: cmd/kube-apiserver/app/server_test.go [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+	network, host string
+}
+
+func TestAuthorizationWebhookDialer(t *testing.T) {
+	tt := []struct {
+		expectedNamespace, expectedName string
+		serverAddress                   string
+		expectedError                   error
+		// If set, passes a nil http.Transport to the function under test.
+		useNilTransport bool
```
Comment (ericchiang) [inline cmd/kube-apiserver/app/server_test.go]: not used?

File: cmd/kube-apiserver/app/server_test.go [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+		fmt.Fprintf(w, "hello")
+	}))
+	defer testServer.Close()
+	resolver := fakeServiceResolver{
+		resolveHostport: testServer.URL,
+	}
+
+	for i, ttt := range tt {
+		var transport *http.Transport
+		defaultTransport := &http.Transport{}
```
Comment (ericchiang) [inline cmd/kube-apiserver/app/server_test.go]: Add a dialer that errors, so if you're not querying a service name you don't actually make a call.

File: pkg/kubeapiserver/authorizer/webhook.go [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+// NewWebhookAuthorizer creates a new kubeapiserver specific webhook
+// authorizer.  configFilePath is the filesystem path to a kubeconfig file that
+// holds the webhook configuration.  cacheAuthorizedTTL is the time for which
+// the positive responses from the authorizer are cached.
+// cacheUnauthorizedTTL is the time for which the negative responses from the
+// authorizer are cached.  dialer is the custom dialer function for establishing
+// the webhook server connections.
+func NewWebhookAuthorizer(
```
Comment (ericchiang) [inline pkg/kubeapiserver/authorizer/webhook.go]: Why was this moved? (sorry if it was brought up previously)

File: cmd/kube-apiserver/app/server.go [hunk 17]
```diff
@@ -444,7 +444,9 @@ func BuildGenericConfig(s *options.ServerRunOptions, proxyTransport *http.Transp
 		return nil, nil, nil, nil, nil, fmt.Errorf("invalid authentication config: %v", err)
 	}
 
-	genericConfig.Authorizer, genericConfig.RuleResolver, err = BuildAuthorizer(s, sharedInformers)
+	genericConfig.Authorizer, genericConfig.RuleResolver, err =
```
Comment (ericchiang) [inline cmd/kube-apiserver/app/server.go]: nit: don't make function calls and definitions span multiple lines like this? other reviewers haven't commented on this, but it's not really how the rest of the code base looks.

**Review by ericchiang — COMMENTED** [review]
have some comments about the added test

File: pkg/kubeapiserver/authorizer/webhook.go (hunk 16 above)
Comment (filmil) [reply pkg/kubeapiserver/authorizer/webhook.go]: It seemed to me to be a good place to put it.  Do you have any actionable feedback for me here?

Comment (ericchiang) [reply pkg/kubeapiserver/authorizer/webhook.go]: Sorry misinterpreted this as moving the entire `webhook.WebhookAuthorizer` carry on.

**Comment by ericchiang** [comment]
One overall comment I have is that I'd be nice to make this accessible to other webhooks (I'm partial to adding this support to the authentication webhook). But I can bring this up in another issue.

**Comment by filmil** [comment]
On Fri, Oct 27, 2017 at 11:10 AM Eric Chiang <notifications@github.com>
wrote:

> One overall comment I have is that I'd be nice to make this accessible to
> other webhooks (I'm partial to adding this support to the authentication
> webhook). But I can bring this up in another issue.
>

Yes, indeed, please bring it up as a separate issue.  I would much prefer
not to expand the scope of this change.

F

**Comment by filmil** [comment]
I'm going to ask @frankfarzan to take this pull request over as I'll be away for the next 2 weeks.  He's got a branch ready with the identical changes at PR #54733

**Comment by liggitt** [comment]
ack, will pick it up there
